cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+            return self._ccnot_true(c0, c1, target)
+
+    def decompose_x(self,
+                    controls: List['cirq.Qid'],
+                    target: 'cirq.Qid',
+                    free_qubits: List['cirq.Qid'] = None):
+        """Implements action of multi-controlled Pauli X gate .
+
+        Result is guaranteed to consist exclusively of 1-qubit, CNOT and CCNOT
+        gates.
+
+        If `free_qubits` has at least 1 element, result has length
+        O(len(controls)).
+
+        Args:
+            controls - control qubits. Can be empty.
+            targets - target qubits.
+            free_qubits - qubits which are neither controlled nor target. Can
+                be modified by algorithm but will end up in inital state.
+        """
+        if free_qubits is None:
+            free_qubits = []
+        m = len(controls)
+        if m == 0:
+            return [ops.X.on(target)]
+        elif m == 1:
+            return [ops.CNOT.on(controls[0], target)]
+        elif m == 2:
+            return self._ccnot(controls[0], controls[1], target)
+
+        m = len(controls)
+        n = m + 1 + len(free_qubits)
+        if (n >= 2 * m - 1) and (m >= 3):
+            # See [1], Lemma 7.2.
+            gates1 = [
+                self._ccnot(controls[m - 2 - i],
+                            free_qubits[m - 4 - i],
+                            free_qubits[m - 3 - i],
+                            congruent=True) for i in range(m - 3)
+            ]
+            gates2 = self._ccnot(controls[0],
+                                 controls[1],
+                                 free_qubits[0],
+                                 congruent=True)
+            gates3 = _flatten(gates1) + gates2 + _flatten(gates1[::-1])
+            first_ccnot = self._ccnot(controls[m - 1], free_qubits[m - 3],
+                                      target)
+            return first_ccnot + gates3 + first_ccnot + gates3
+        elif len(free_qubits) >= 1:
+            # See [1], Lemma 7.3.
+            m1 = n // 2
+            free1 = controls[m1:] + [target] + free_qubits[1:]
+            ctrl1 = controls[:m1]
+            gates1 = self.decompose_x(ctrl1, free_qubits[0], free_qubits=free1)
+            free2 = controls[:m1] + free_qubits[1:]
+            ctrl2 = controls[m1:] + [free_qubits[0]]
+            gates2 = self.decompose_x(ctrl2, target, free_qubits=free2)
+            return gates1 + gates2 + gates1 + gates2
+        else:
+            # No free qubit - must use main algorithm.
+            # This will never happen if called from main algorithm and is added
+            # only for completeness.
+            X = np.array([[0, 1], [1, 0]])
+            return self.decompose(X, controls, target)
+
+    def decompose(self,
+                  matrix: np.ndarray,
+                  controls: List['cirq.Qid'],
+                  target: 'cirq.Qid',
+                  free_qubits: List['cirq.Qid'] = None,
+                  power: float = 1.0):
```

Comment:
Drop the `power` argument. If the user wants a different operation (such as a smaller power), they should just provide the matrix for that operation.

Reply:
I added this to prevent precision loss.
I agree it doesn't add value to user, but I call this method recursively, and every time I call it I need to calculate square root of a matrix. I found that it's more accurate to pass the same matrix but device power by 2.

If I make this method private, and provide public method without "power" argument - will it solve this problem?

Reply:
Removed `power` from public function, but I still left it in the `_decompose_recursive` function, because it improves precision.